src/plugins/intel_gpu/src/graph/graph_optimizer/post_optimize_weights.cpp

```diff
@@ -38,53 +40,81 @@ void post_optimize_weights::optimize_weights(T& node, program& p) {
-            // set generic_layer's node output layout and implementation
-            auto& g_node = node.get_dependency(i);
-            g_node.get_output_layout(false);
-
-            // Don't run impl selection to avoid double compilation of reorder kernels
-            // in main program and internal program for constant propagation
-            if ((!g_node.is_constant()) && (!reorder.second)) {
-                g_node.set_selected_impl(g_node.type()->choose_impl(g_node));
-                if (auto impl = g_node.get_selected_impl()) {
-                    auto params = g_node.get_kernel_impl_params();
-                    p.get_kernels_cache().add_kernels_source(*params, impl->get_kernels_source());
+        program_node& prev_node = node.get_dependency(i);
+
+        if (weights_reorder_params != nullptr) {
+            bool can_be_fused = prev_node.is_type<reorder>() &&
+                                prev_node.get_users().size() == 1 &&
+                                prev_node.get_dependencies().size() == 1 &&
+                                !prev_node.has_fused_primitives() &&
+                                !prev_node.as<reorder>().has_mean() &&
+                                prev_node.as<reorder>().get_primitive()->subtract_per_feature.empty();
+            if (can_be_fused) {
+                data_types input_dtype = prev_node.get_input_layouts()[0].data_type;
+                auto updated_input_layout = weights_reorder_params->get_input_layout();
+                updated_input_layout.data_type = input_dtype;
+                weights_reorder_params->set_input_layout(updated_input_layout);
+
+                auto weights_reorder = _rf.get_weights_reorder(prev_node.get_primitive()->input[0].pid,
+                                                               weights_reorder_params);
+                auto& weights_reorder_node = p.get_or_create(weights_reorder.first);
+                p.replace(prev_node, weights_reorder_node);
+                weights_reorder_node.recalc_output_layout(false);
+
+                if (!weights_reorder.second) {
+                    set_implementation(weights_reorder_node);
+                }
+            } else {
+                auto weights_reorder = _rf.get_weights_reorder(prev_node.id(), weights_reorder_params);
+                // insert new weights reorder node to topology
+                p.add_intermediate(weights_reorder.first, node, i, !weights_reorder.second);
+                // set weights reorder's node output layout and implementation
+                auto& weights_reorder_node = node.get_dependency(i);
+                weights_reorder_node.get_output_layout(false);
+
+                if (!weights_reorder.second) {
+                    set_implementation(weights_reorder_node);
                 }
             }
         }
     }
-
-    // Reset weights reorder params to not keep source code pointer
-    impl->reset_weights_reorder_params();
```

Comment:
I think here was the only usage of `reset_weights_reorder_params`, so the method can be removed as well.